src/esp32/dtls.rs

```diff
@@ -0,0 +1,755 @@
+            mbedtls_ssl_config_init(self.ssl_config.as_mut());
+            mbedtls_x509_crt_init(self.x509.as_mut());
+            mbedtls_pk_init(self.pk_ctx.as_mut());
+            mbedtls_entropy_init(self.dtls_entropy.as_mut());
+            mbedtls_ctr_drbg_init(self.drbg_ctx.as_mut());
+        }
+        let ret = unsafe {
+            //TODO(npm) we can avoid an allocation if we use the nocpy version
+            mbedtls_x509_crt_parse_der(
+                self.x509.as_mut(),
+                certificate.get_der_certificate().as_ptr(),
+                certificate.get_der_certificate().len(),
+            )
+        };
+        if ret != 0 {
+            return Err(SSLError::SSLCertParseFail(ret));
+        }
+        let ret = unsafe {
+            mbedtls_pk_parse_key(
+                self.pk_ctx.as_mut(),
+                certificate.get_der_keypair().as_ptr(),
+                certificate.get_der_keypair().len(),
+                std::ptr::null(),
+                0,
+            )
+        };
+        if ret != 0 {
+            return Err(SSLError::SSLKeyParseFail(ret));
+        }
+
+        let ret = unsafe {
+            mbedtls_ctr_drbg_seed(
+                self.drbg_ctx.as_mut(),
+                Some(mbedtls_entropy_func),
+                self.dtls_entropy.as_mut() as *mut mbedtls_entropy_context as *mut _,
+                std::ptr::null(),
+                0,
+            )
+        };
+        if ret != 0 {
+            return Err(SSLError::SSLEntropySeedFailure(ret));
+        }
+
+        let ret = unsafe {
+            mbedtls_ssl_config_defaults(
+                self.ssl_config.as_mut(),
+                MBEDTLS_SSL_IS_SERVER as i32,
+                MBEDTLS_SSL_TRANSPORT_DATAGRAM as i32,
+                MBEDTLS_SSL_PRESET_DEFAULT as i32,
+            )
+        };
+        if ret != 0 {
+            return Err(SSLError::SSLConfigFailure(ret));
+        }
+        unsafe {
+            mbedtls_ssl_conf_rng(
+                self.ssl_config.as_mut(),
+                Some(mbedtls_ctr_drbg_random),
+                self.drbg_ctx.as_mut() as *mut mbedtls_ctr_drbg_context as *mut c_void,
+            );
+            // if wee need to debug the handshake
+            mbedtls_debug_set_threshold(0);
+            mbedtls_ssl_conf_dbg(
+                self.ssl_config.as_mut(),
+                Some(ssl_debug),
+                std::ptr::null_mut(),
+            );
+
+            // Cookie are disabled, we might want to re-enable them at a later stage or only accept ClientHello originating from
+            // the selected pair (or any DTLS packets)
+            // see 4.2.1.  Denial-of-Service Countermeasures (RFC 6347)
+            mbedtls_ssl_conf_dtls_cookies(
+                self.ssl_config.as_mut(),
+                None,
+                None,
+                std::ptr::null_mut(),
+            );
+
+            //(TODO(npm)) Attempt to weak link mbedtls_ssl_conf_dbg
```

Comment:
must this be done?

Reply:
it can be useful if we need to debug DTLS negotiation on the ESP32 side but we can add it on a ad-hoc basis